File: benchmarks/tsb/bench_ffill_bfill_series_na.ts
/**
 * Benchmark: ffillSeries / bfillSeries — forward/backward fill on 100k-element Series.
 * Outputs JSON: {"function": "ffill_bfill_series_na", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, ffillSeries, bfillSeries } from "../../src/index.ts";

const SIZE = 100_000;
const WARMUP = 5;
const ITERATIONS = 50;

// ~10% null values
const data: (number | null)[] = Array.from({ length: SIZE }, (_, i) =>
  i % 10 === 0 ? null : i * 1.5,
);
const s = new Series({ data });

for (let i = 0; i < WARMUP; i++) {
  ffillSeries(s);
  bfillSeries(s);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  ffillSeries(s);
  bfillSeries(s);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "ffill_bfill_series_na",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_ffill_bfill_series_na.py
"""
Benchmark: Series.ffill() / bfill() — forward/backward fill on 100k-element Series.
Outputs JSON: {"function": "ffill_bfill_series_na", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
WARMUP = 5
ITERATIONS = 50

data = np.arange(SIZE, dtype=float) * 1.5
data[::10] = np.nan
s = pd.Series(data)

for _ in range(WARMUP):
    s.ffill()
    s.bfill()

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.ffill()
    s.bfill()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "ffill_bfill_series_na",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
